### ISSUE TITLE
Random race condition in test workflows (in package internal/command)

### ISSUE THREAD

**Issue by StephanHCB** [opentofu/opentofu#1144]
### OpenTofu Version

```shell
commit fcd241c0e4 (mainline from a few days ago)
```

### OpenTofu Configuration Files

n/a

### Debug Output

```
==================
WARNING: DATA RACE
Write at 0x00c00032e670 by goroutine 6749:
  github.com/opentofu/opentofu/internal/tofu.(*MockProvider).ConfigureProvider()
      /home/runner/work/opentofu/opentofu/internal/tofu/provider_mock.go:247 +0xe4
  github.com/opentofu/opentofu/internal/tofu.(*BuiltinEvalContext).ConfigureProvider()
      /home/runner/work/opentofu/opentofu/internal/tofu/eval_context_builtin.go:204 +0x430
  github.com/opentofu/opentofu/internal/tofu.(*NodeApplyableProvider).ConfigureProvider()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_provider.go:166 +0x1141
  github.com/opentofu/opentofu/internal/tofu.(*NodeApplyableProvider).Execute()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_provider.go:45 +0x7a8
  github.com/opentofu/opentofu/internal/tofu.(*ContextGraphWalker).Execute()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph_walk_context.go:143 +0x159
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk.func1()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:78 +0x38f
  github.com/opentofu/opentofu/internal/dag.(*Walker).walkVertex()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:384 +0x426
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update.func3()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x5d

Previous read at 0x00c00032e670 by goroutine 6745:
  github.com/opentofu/opentofu/internal/tofu.(*MockProvider).ApplyResourceChange()
      /home/runner/work/opentofu/opentofu/internal/tofu/provider_mock.go:402 +0x12f
  github.com/opentofu/opentofu/internal/tofu.(*NodeAbstractResourceInstance).apply()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_resource_abstract_instance.go:2296 +0x17fc
  github.com/opentofu/opentofu/internal/tofu.(*NodeDestroyResourceInstance).managedResourceExecute()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_resource_destroy.go:217 +0xc7d
  github.com/opentofu/opentofu/internal/tofu.(*NodeDestroyResourceInstance).Execute()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_resource_destroy.go:144 +0xea
  github.com/opentofu/opentofu/internal/tofu.(*ContextGraphWalker).Execute()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph_walk_context.go:143 +0x159
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk.func1()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:78 +0x38f
  github.com/opentofu/opentofu/internal/dag.(*Walker).walkVertex()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:384 +0x426
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update.func3()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x5d

Goroutine 6749 (running) created at:
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x1384
  github.com/opentofu/opentofu/internal/dag.(*AcyclicGraph).Walk()
      /home/runner/work/opentofu/opentofu/internal/dag/dag.go:167 +0xe8
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:137 +0xfb
  github.com/opentofu/opentofu/internal/tofu.(*Graph).Walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:38 +0x138
  github.com/opentofu/opentofu/internal/tofu.(*Context).walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/context_walk.go:56 +0x122
  github.com/opentofu/opentofu/internal/tofu.(*Context).Apply()
      /home/runner/work/opentofu/opentofu/internal/tofu/context_apply.go:61 +0x6e6
  github.com/opentofu/opentofu/internal/command.(*TestFileRunner).apply.func1()
      /home/runner/work/opentofu/opentofu/internal/command/test.go:814 +0x224

Goroutine 6745 (running) created at:
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x1384
  github.com/opentofu/opentofu/internal/dag.(*AcyclicGraph).Walk()
      /home/runner/work/opentofu/opentofu/internal/dag/dag.go:167 +0xe8
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:137 +0xfb
  github.com/opentofu/opentofu/internal/tofu.(*Graph).Walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:38 +0x138
  github.com/opentofu/opentofu/internal/tofu.(*Context).walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/context_walk.go:56 +0x122
  github.com/opentofu/opentofu/internal/tofu.(*Context).Apply()
      /home/runner/work/opentofu/opentofu/internal/tofu/context_apply.go:61 +0x6e6
  github.com/opentofu/opentofu/internal/command.(*TestFileRunner).apply.func1()
      /home/runner/work/opentofu/opentofu/internal/command/test.go:814 +0x224
==================
==================
WARNING: DATA RACE
Write at 0x00c000cc1d68 by goroutine 6749:
  github.com/opentofu/opentofu/internal/command/testing.(*TestProvider).ConfigureProvider()
      /home/runner/work/opentofu/opentofu/internal/command/testing/test_provider.go:173 +0xb7
  github.com/opentofu/opentofu/internal/command/testing.(*TestProvider).ConfigureProvider-fm()
      <autogenerated>:1 +0xc4
  github.com/opentofu/opentofu/internal/tofu.(*MockProvider).ConfigureProvider()
      /home/runner/work/opentofu/opentofu/internal/tofu/provider_mock.go:251 +0x2f8
  github.com/opentofu/opentofu/internal/tofu.(*BuiltinEvalContext).ConfigureProvider()
      /home/runner/work/opentofu/opentofu/internal/tofu/eval_context_builtin.go:204 +0x430
  github.com/opentofu/opentofu/internal/tofu.(*NodeApplyableProvider).ConfigureProvider()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_provider.go:166 +0x1141
  github.com/opentofu/opentofu/internal/tofu.(*NodeApplyableProvider).Execute()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_provider.go:45 +0x7a8
  github.com/opentofu/opentofu/internal/tofu.(*ContextGraphWalker).Execute()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph_walk_context.go:143 +0x159
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk.func1()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:78 +0x38f
  github.com/opentofu/opentofu/internal/dag.(*Walker).walkVertex()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:384 +0x426
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update.func3()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x5d

Previous read at 0x00c000cc1d68 by goroutine 6745:
  github.com/opentofu/opentofu/internal/command/testing.(*TestProvider).GetResourceKey()
      /home/runner/work/opentofu/opentofu/internal/command/testing/test_provider.go:120 +0x47
  github.com/opentofu/opentofu/internal/command/testing.(*TestProvider).ApplyResourceChange()
      /home/runner/work/opentofu/opentofu/internal/command/testing/test_provider.go:201 +0x43c
  github.com/opentofu/opentofu/internal/command/testing.(*TestProvider).ApplyResourceChange-fm()
      <autogenerated>:1 +0xab
  github.com/opentofu/opentofu/internal/tofu.(*MockProvider).ApplyResourceChange()
      /home/runner/work/opentofu/opentofu/internal/tofu/provider_mock.go:408 +0x454
  github.com/opentofu/opentofu/internal/tofu.(*NodeAbstractResourceInstance).apply()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_resource_abstract_instance.go:2296 +0x17fc
  github.com/opentofu/opentofu/internal/tofu.(*NodeDestroyResourceInstance).managedResourceExecute()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_resource_destroy.go:217 +0xc7d
  github.com/opentofu/opentofu/internal/tofu.(*NodeDestroyResourceInstance).Execute()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_resource_destroy.go:144 +0xea
  github.com/opentofu/opentofu/internal/tofu.(*ContextGraphWalker).Execute()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph_walk_context.go:143 +0x159
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk.func1()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:78 +0x38f
  github.com/opentofu/opentofu/internal/dag.(*Walker).walkVertex()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:384 +0x426
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update.func3()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x5d

Goroutine 6749 (running) created at:
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x1384
  github.com/opentofu/opentofu/internal/dag.(*AcyclicGraph).Walk()
      /home/runner/work/opentofu/opentofu/internal/dag/dag.go:167 +0xe8
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:137 +0xfb
  github.com/opentofu/opentofu/internal/tofu.(*Graph).Walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:38 +0x138
  github.com/opentofu/opentofu/internal/tofu.(*Context).walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/context_walk.go:56 +0x122
  github.com/opentofu/opentofu/internal/tofu.(*Context).Apply()
      /home/runner/work/opentofu/opentofu/internal/tofu/context_apply.go:61 +0x6e6
  github.com/opentofu/opentofu/internal/command.(*TestFileRunner).apply.func1()
      /home/runner/work/opentofu/opentofu/internal/command/test.go:814 +0x224

Goroutine 6745 (running) created at:
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x1384
  github.com/opentofu/opentofu/internal/dag.(*AcyclicGraph).Walk()
      /home/runner/work/opentofu/opentofu/internal/dag/dag.go:167 +0xe8
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:137 +0xfb
  github.com/opentofu/opentofu/internal/tofu.(*Graph).Walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:38 +0x138
  github.com/opentofu/opentofu/internal/tofu.(*Context).walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/context_walk.go:56 +0x122
  github.com/opentofu/opentofu/internal/tofu.(*Context).Apply()
      /home/runner/work/opentofu/opentofu/internal/tofu/context_apply.go:61 +0x6e6
  github.com/opentofu/opentofu/internal/command.(*TestFileRunner).apply.func1()
      /home/runner/work/opentofu/opentofu/internal/command/test.go:814 +0x224
==================
--- FAIL: TestTest (0.90s)
    --- FAIL: TestTest/no_providers_in_main (0.07s)
        testing.go:1465: race detected during execution of test
    testing.go:1465: race detected during execution of test
FAIL
FAIL	github.com/opentofu/opentofu/internal/command	56.954s
```

### Expected Behavior

The github workflow for detecting races does not randomly fail

### Actual Behavior

The github workflow for detecting races randomly sometimes fails for package "internal/command".

Output details see debug output.

```
--- FAIL: TestTest (0.90s)
    --- FAIL: TestTest/no_providers_in_main (0.07s)
        testing.go:1465: race detected during execution of test
    testing.go:1465: race detected during execution of test
FAIL
FAIL	github.com/opentofu/opentofu/internal/command	56.954s
```

### Steps to Reproduce

run the workflows often enough (happened on my PR a moment ago)

### Additional Context

Tried to reproduce on current main, didn't happen but then it's random, so I can't say whether it was already fixed.

The debug output contains stack traces of every goroutine. Hoping that this can help you find and fix the issue.

**Comment by cube2222** [comment]
Thanks for creating the issue @StephanHCB!

I've hit this too, we should investigate deeper and fix it.

**Comment by desmaraisp** [comment]
I can confirm I was able to reproduce the error with the following workflow:

```
name: Quick Checks (race)

on:
  workflow_dispatch: 

jobs:
  race-tests:
    strategy:
      matrix:
        index: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
    
    name: "Race Tests"
    runs-on: ubuntu-latest

    steps:
      - name: "Fetch source code"
        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3

      - name: Determine Go version
        id: go
        uses: ./.github/actions/go-version

      - name: Install Go toolchain
        uses: actions/setup-go@fac708d6674e30b6ba41289acaab6d4b75aa0753 # v4.0.1
        with:
          go-version: ${{ steps.go.outputs.version }}

      # The race detector add significant time to the unit tests, so only run
      # it for select packages.
      - name: "Race detector"
        run: |
          go test -race ./internal/tofu ./internal/command ./internal/states
```

Even with 20 parallel execution, I can only manage to trigger the issue about ~50% of the time.

Interestingly enough, the data race I got was slightly different from the one reported by @StephanHCB, and I wasn't able to reproduce the exact same issue as the initial message. I think they're the same issue, but it's quite possible they're separate

```
WARNING: DATA RACE
Write at 0x00c000517310 by goroutine 14023:
  github.com/opentofu/opentofu/internal/tofu.(*MockProvider).ImportResourceState()
      /home/runner/work/opentofu/opentofu/internal/tofu/provider_mock.go:483 +0x87a
  github.com/opentofu/opentofu/internal/tofu.(*NodePlannableResourceInstance).importState()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_resource_plan_instance.go:4[6](https://github.com/desmaraisp/opentofu/actions/runs/7635776092/job/20801849203#step:5:7)6 +0x3e4
  github.com/opentofu/opentofu/internal/tofu.(*NodePlannableResourceInstance).managedResourceExecute()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_resource_plan_instance.go:201 +0x924
  github.com/opentofu/opentofu/internal/tofu.(*NodePlannableResourceInstance).Execute()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_resource_plan_instance.go:[7](https://github.com/desmaraisp/opentofu/actions/runs/7635776092/job/20801849203#step:5:8)4 +0xea
  github.com/opentofu/opentofu/internal/tofu.(*ContextGraphWalker).Execute()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph_walk_context.go:143 +0x159
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk.func1()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:7[8](https://github.com/desmaraisp/opentofu/actions/runs/7635776092/job/20801849203#step:5:9) +0x38f
  github.com/opentofu/opentofu/internal/dag.(*Walker).walkVertex()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:384 +0x426
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update.func3()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x5d

Previous read at 0x00c000517310 by goroutine 14020:
  github.com/opentofu/opentofu/internal/tofu.(*NodePlannableResourceInstance).importState()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_resource_plan_instance.go:514 +0xce4
  github.com/opentofu/opentofu/internal/tofu.(*NodePlannableResourceInstance).managedResourceExecute()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_resource_plan_instance.go:201 +0x[9](https://github.com/desmaraisp/opentofu/actions/runs/7635776092/job/20801849203#step:5:10)24
  github.com/opentofu/opentofu/internal/tofu.(*NodePlannableResourceInstance).Execute()
      /home/runner/work/opentofu/opentofu/internal/tofu/node_resource_plan_instance.go:74 +0xea
  github.com/opentofu/opentofu/internal/tofu.(*ContextGraphWalker).Execute()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph_walk_context.go:143 +0x159
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk.func1()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:78 +0x38f
  github.com/opentofu/opentofu/internal/dag.(*Walker).walkVertex()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:384 +0x426
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update.func3()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x5d

Goroutine 14023 (running) created at:
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x1384
  github.com/opentofu/opentofu/internal/dag.(*AcyclicGraph).Walk()
      /home/runner/work/opentofu/opentofu/internal/dag/dag.go:167 +0xe8
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:137 +0xfb
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk.func1()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:[11](https://github.com/desmaraisp/opentofu/actions/runs/7635776092/job/20801849203#step:5:12)9 +0x91e
  github.com/opentofu/opentofu/internal/dag.(*Walker).walkVertex()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:384 +0x426
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update.func3()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x5d

Goroutine 14020 (running) created at:
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x[13](https://github.com/desmaraisp/opentofu/actions/runs/7635776092/job/20801849203#step:5:14)84
  github.com/opentofu/opentofu/internal/dag.(*AcyclicGraph).Walk()
      /home/runner/work/opentofu/opentofu/internal/dag/dag.go:167 +0xe8
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:137 +0xfb
  github.com/opentofu/opentofu/internal/tofu.(*Graph).walk.func1()
      /home/runner/work/opentofu/opentofu/internal/tofu/graph.go:119 +0x91e
  github.com/opentofu/opentofu/internal/dag.(*Walker).walkVertex()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:384 +0x426
  github.com/opentofu/opentofu/internal/dag.(*Walker).Update.func3()
      /home/runner/work/opentofu/opentofu/internal/dag/walk.go:307 +0x5d
==================
--- FAIL: TestContext2Plan_importIntoModuleWithGeneratedConfig (0.03s)
    testing.go:[14](https://github.com/desmaraisp/opentofu/actions/runs/7635776092/job/20801849203#step:5:15)65: race detected during execution of test
FAIL
FAIL	github.com/opentofu/opentofu/internal/tofu	[33](https://github.com/desmaraisp/opentofu/actions/runs/7635776092/job/20801849203#step:5:34).0[35](https://github.com/desmaraisp/opentofu/actions/runs/7635776092/job/20801849203#step:5:36)s
ok  	github.com/opentofu/opentofu/internal/command	58.[40](https://github.com/desmaraisp/opentofu/actions/runs/7635776092/job/20801849203#step:5:41)3s
ok  	github.com/opentofu/opentofu/internal/states	1.0[52](https://github.com/desmaraisp/opentofu/actions/runs/7635776092/job/20801849203#step:5:53)s
FAIL
```

I'll have to say though, I know absolutely nothing of go (first time touching it), but from what I could parse this is related to initializing p.ImportResourceStateResponse [here](https://github.com/opentofu/opentofu/blob/8f23c1a8d2852fb1e70fff7aa66e50d538d3de8a/internal/tofu/context_plan2_test.go#L5177). From what I can see, it seems like multiple goroutines are being started somewhere in managedResourceExecute(). Since the same slice is shared between those goroutines, we end up [reading](https://github.com/opentofu/opentofu/blob/8f23c1a8d2852fb1e70fff7aa66e50d538d3de8a/internal/tofu/node_resource_plan_instance.go#L475) and [writing](https://github.com/opentofu/opentofu/blob/8f23c1a8d2852fb1e70fff7aa66e50d538d3de8a/internal/tofu/provider_mock.go#L483) to the slice at the same time (pointer shenanigans).

I've forked and prototyped a possible fix over at #1182. I've run the race tests by batches of 20 a couple times, and the data race warning no longer seems to pop up whereas it was showing up somewhat consistently before the fix.

**Comment by desmaraisp** [comment]
Thank you! I've marked the PR as ready to review
